Fix service restoration and extend CI for externalTrafficPolicy=Local services

gandro opened the pull request:
This PR extends the CI tests for `externalTrafficPolicy=Local` services and fixes a bug in the service cache restoration unconvered by said tests.

The new CI tests deploy a `NodePort` service with `externalTrafficPolicy=Local` where only a single node is running a service endpoint. It tests that only direct connections to that node will succeed, while any requests to a node without a local backend are dropped.

While the test itself worked as expected, the service entries on nodes without any backends caused subsequent Cilium CI status preflight checks to fail, as the new services were not garbage collected by cilium-agent after a restart. Commit 21e528037ac092de8f653d0c25b06dbf49abf653 fixes this issue by ensuring that backend-less services are also restored from the datapath. This allows cilium-agent to subsequently garbage collect these entries once it recognizes that these services have been removed from Kubernetes.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9778)
<!-- Reviewable:end -->

brb reviewed (approved):
Nice catch and thanks for the tests!

gandro committed:
lbmap: Include services without backends in DumpServiceMaps

This changes the behavior of DumpServiceMaps to include services without
any backends. This is required when DumpServiceMaps is used to restore
services in the service cache. Without this change, services without any
backends (e.g. NodePort services with externalTrafficPolicy=Local) will
not be restored from the datapath and are consequently not garbage
collected once cilium-agent recognizes that these services have been
removed from Kubernetes.

This commit therefore fixes a case where the Cilium CI status preflight
check fails to pass, as the lbmap contains such leftover entries.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
test: Extend externalTrafficPolicy=Local tests

Adds an additional check for NodePort services with externalTrafficPolicy=Local
where only a single node is running a service endpoint. Therefore, we test that
only direct connections to that node will succeed, while any requests to a node
without a local backend are dropped.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
